File: services/api/internal/domain/media/presenter.go
// Package media owns station programming and replay rules.
package media

import (
	"errors"
	"strings"

	"github.com/kurasikapa/api/internal/domain/identity"
	"github.com/kurasikapa/api/internal/domain/shared"
)

var ErrEmptyPresenterName = errors.New("presenter name cannot be empty")

type PresenterState struct {
	ID                                  shared.PresenterID
	Name, Slug, Locale, Role, Biography string
	PortraitAssetID                     *shared.AssetID
	Published                           bool
	CreatedBy                           shared.UserID
}

type Presenter struct{ state PresenterState }

func NewPresenter(actor identity.Actor, state PresenterState) (Presenter, error) {
	if err := actor.Require(identity.PermStreamManage); err != nil {
		return Presenter{}, err
	}
	state.Name = strings.TrimSpace(state.Name)
	if state.Name == "" {
		return Presenter{}, ErrEmptyPresenterName
	}
	state.Published = false
	state.CreatedBy = actor.ID()
	return Presenter{state: state}, nil
}

func ReconstitutePresenter(state PresenterState) Presenter { return Presenter{state: state} }
func (p Presenter) State() PresenterState                  { return p.state }
func (p Presenter) ID() shared.PresenterID                 { return p.state.ID }
func (p Presenter) Publish(actor identity.Actor) (Presenter, error) {
	if err := actor.Require(identity.PermStreamManage); err != nil {
		return Presenter{}, err
	}
	p.state.Published = true
	return p, nil
}

File: packages/domain/src/media/presenter.ts
import { type Actor, requirePermission } from '../identity/actor'
import type { AssetId, PresenterId, UserId } from '../shared/ids'

export interface PresenterProps {
  readonly id: PresenterId
  readonly name: string
  readonly slug: string
  readonly locale: string
  readonly role: string
  readonly biography: string
  readonly portraitAssetId: AssetId | null
  readonly published: boolean
  readonly createdBy: UserId
}

export type NewPresenter = Omit<PresenterProps, 'published' | 'createdBy'>

export class EmptyPresenterName extends Error {
  constructor() {
    super('Presenter name cannot be empty')
    this.name = 'EmptyPresenterName'
  }
}

/** A station-owned public identity used by programmes, schedules and bylines. */
export class Presenter {
  private constructor(private readonly props: PresenterProps) {}

  static create(actor: Actor, input: NewPresenter): Presenter {
    requirePermission(actor, 'stream:manage')
    if (input.name.trim() === '') throw new EmptyPresenterName()

    return new Presenter({ ...input, name: input.name.trim(), published: false, createdBy: actor.id })
  }

  static reconstitute(props: PresenterProps): Presenter {
    return new Presenter(props)
  }

  get id(): PresenterId { return this.props.id }
  get name(): string { return this.props.name }
  get slug(): string { return this.props.slug }
  get locale(): string { return this.props.locale }
  get role(): string { return this.props.role }
  get biography(): string { return this.props.biography }
  get portraitAssetId(): AssetId | null { return this.props.portraitAssetId }
  get published(): boolean { return this.props.published }

  publish(actor: Actor): Presenter {
    requirePermission(actor, 'stream:manage')
    return new Presenter({ ...this.props, published: true })
  }

  snapshot(): PresenterProps {
    return this.props
  }
}
